src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,218 @@
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody))
+                {
+                    originalBodyFeature = context.Features.Get<IHttpResponseBodyFeature>()!;
+                    // TODO pool these.
+                    responseBufferingStream = new ResponseBufferingStream(originalBodyFeature,
+                        options.ResponseBodyLogLimit,
+                        _logger,
+                        context,
+                        options.SupportedMediaTypes);
+                    response.Body = responseBufferingStream;
+                }
+
+                await _next(context).ConfigureAwait(false);
+                var list = new List<KeyValuePair<string, object?>>(
+                    response.Headers.Count + DefaultResponseFieldsMinusHeaders);
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponseStatusCode))
+                {
+                    list.Add(new KeyValuePair<string, object?>(nameof(response.StatusCode), response.StatusCode));
+                }
+
+                if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponseHeaders))
+                {
+                    FilterHeaders(list, response.Headers, options.ResponseHeaders);
+                }
+
+                var httpResponseLog = new HttpResponseLog(list);
+
+                _logger.Log(LogLevel.Information,
+                    eventId: LoggerEventIds.ResponseLog,
+                    state: httpResponseLog,
+                    exception: null,
+                    formatter: HttpResponseLog.Callback);
+
+                if (responseBufferingStream != null)
+                {
+                    responseBufferingStream.LogString(responseBufferingStream.Encoding, LoggerEventIds.ResponseBody, CoreStrings.ResponseBody);
+                }
+            }
+            finally
+            {
+                responseBufferingStream?.Dispose();
+
+                if (originalBodyFeature != null)
+                {
+                    context.Features.Set(originalBodyFeature);
+                }
+
+                requestBufferingStream?.Dispose();
+
+                if (originalBody != null)
+                {
+                    context.Request.Body = originalBody;
+                }
+            }
+        }
+
+        private static void AddToList(List<KeyValuePair<string, object?>> list, string key, string? value)
+        {
+            list.Add(new KeyValuePair<string, object?>(key, value));
+        }
+
+        private void FilterHeaders(List<KeyValuePair<string, object?>> keyValues,
+            IHeaderDictionary headers,
+            ISet<string> allowedHeaders)
+        {
+            foreach (var (key, value) in headers)
+            {
+                if (!allowedHeaders.Contains(key))
+                {
+                    // Key is not among the "only listed" headers.
+                    keyValues.Add(new KeyValuePair<string, object?>(key, "X"));
```

Comment:
Add a private const field with "X".
1. Easier to understand what is going on when it has a name
2. Easier to change if we want a different value